peer.go

```diff
@@ -0,0 +1,1087 @@
+	// term, the local server will update and become a follower.
+	term uint64
+
+	// Set to true if the peer granted this server a vote in 'term'.
+	voteGranted bool
+
+	// The index in the log that the remote server has acknowledged as matching
+	// this server's. Reset to 0 when the term changes.
+	matchIndex uint64
+
+	// A lower bound of when the peer last heard from us. Upon the receipt of any
+	// reply from the peer (successful or not), this is updated to the time the
+	// request was started.
+	lastContact time.Time
+
+	// An upper bound of when the peer last heard from us. Upon the receipt of any
+	// reply from the peer (successful or not), this is updated to the current
+	// time.
+	lastReply time.Time
+
+	// The value of control.verifyCounter when the last completed heartbeat to the
+	// peer was sent.
+	verifiedCounter uint64
+}
+
+// The main type of a Peer (peer clashes too much), which is in charge of all
+// communication with a single remote server. Peers can span many terms and are
+// normally kept until the remote server is removed from the cluster or the
+// local server is shutdown.
+type peerState struct {
+	// This field is shared with other Peer goroutines.
+	shared *peerShared
+
+	///// The rest of this stuff is private to the main Peer goroutine. /////
+
+	// The Raft module sends on this channel whenever any of the 'control' fields
+	// change (unbuffered).
+	controlCh <-chan peerControl
+
+	// The Peer module sends on this channel to the Raft module shortly after any
+	// of the 'progress' fields change (unbuffered).
+	progressCh chan<- peerProgress
+
+	// The latest control information received from the Raft module.
+	control peerControl
+
+	// The latest progress information computed by the Peer.
+	progress peerProgress
+
+	// If true, the Raft module has not yet received the latest updates to
+	// 'progress', and these will be sent at the first availble opportunity.
+	sendProgress bool
+
+	// If true, a vote has already been requested from this Peer during
+	// control.term.
+	sentRequestVote bool
+
+	// The last time a heartbeat has been sent to this Peer during control.term.
+	// As leader, this will send another heartbeat after
+	// lastHeartbeatSent + heartbeatInterval.
+	lastHeartbeatSent time.Time
+
+	// As leader, nextIndex is the index of the next log entry to send to the
+	// peer, which may fall one entry past the end of the log.
+	nextIndex uint64
+
+	// allowPipeline is true when AppendEntries RPCs should be sent over a
+	// pipeline (a separate path through the Transport), false otherwise.
+	allowPipeline bool
+
+	// If true, the peer needs a snapshot because this server has already
+	// discarded the log entries it would otherwise send in AppendEntries.
+	needsSnapshot bool
+
+	// Set to true while an InstallSnapshot RPC is in progress. This prevents
+	// further InstallSnapshot and bulk AppendEntries RPCs from starting.
+	outstandingInstallSnapshotRPC bool
```

Comment:
These few "outstanding*" fields should probably be per term (cleared out when the term changes and only decremented if the term hasn't changed). Otherwise after a term switch, you might needlessly wait around for the old ones to finish/timeout.